zrml/swaps/src/tests.rs

```diff
@@ -1569,10 +1642,63 @@ fn swap_exact_amount_out_exchanges_correct_values_with_cpmm(
     });
 }
 
+#[test]
+fn swap_exact_amount_out_exchanges_correct_values_with_cpmm_with_fees() {
+    ExtBuilder::default().build().execute_with(|| {
+        frame_system::Pallet::<Runtime>::set_block_number(1);
+        let asset_amount_out = _1;
+        let asset_amount_in = 11223344556; // 10101010100 / 0.9
+        let asset_bound = Some(_2);
+        let max_price = Some(_3);
+        frame_system::Pallet::<Runtime>::set_block_number(1);
+        ASSETS.iter().cloned().for_each(|asset| {
+            let _ = Currencies::deposit(asset, &ALICE, _25);
+            let _ = Currencies::deposit(asset, &BOB, _10000);
+        });
+        assert_ok!(Swaps::create_pool(
+            BOB,
+            ASSETS.iter().cloned().collect(),
+            ASSETS.last().unwrap().clone(),
+            0,
+            ScoringRule::CPMM,
+            Some(BASE / 10),
+            Some(<Runtime as crate::Config>::MinLiquidity::get()),
+            Some(vec!(_2, _2, _2, _2)),
+        ));
+        assert_ok!(Swaps::swap_exact_amount_out(
+            alice_signed(),
+            0,
+            ASSET_A,
+            asset_bound,
+            ASSET_B,
+            asset_amount_out,
+            max_price,
+        ));
+        System::assert_last_event(
+            Event::SwapExactAmountOut(SwapEvent {
+                asset_amount_in,
+                asset_amount_out,
+                asset_bound,
+                asset_in: ASSET_A,
+                asset_out: ASSET_B,
+                cpep: CommonPoolEventParams { pool_id: 0, who: 0 },
+                max_price,
+            })
+            .into(),
+        );
+        assert_all_parameters(
+            [_25 - asset_amount_in, _25 + asset_amount_out, _25, _25],
+            0,
+            [_100 + asset_amount_in, _100 - asset_amount_out, _100, _100],
+            _100,
+        );
+    });
+}
```

Comment:
>  I would expect, that every time there is a trade (e. g. `swap_exact_amount_out`) on the pool, the trader has to pay the swap fee amount of `0.1 * BASE`.

No, everytime there is a trade, Alice pays 10% of the tokens that she's swapping in as trading fee.

> So the idea with `assert_all_parameters` is to show, that Alice did pay the fee with the `asset_in`, because `asset_amount_out` is `BASE` and `asset_amount_in` is `0.1 * BASE` greater than BASE (so `1.1 * BASE`)?

It's actually `1.111... * BASE` (` = BASE / 0.9`), but other than that, yes.